src/EditorFeatures/Core/Tagging/AggregateTagger.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.CodeAnalysis.Editor.Tagging;
+
+/// <summary>
+/// Simple tagger that aggregates the underlying syntax/semantic compiler/analyzer taggers and presents them as
+/// a single event source and source of tags.
+/// </summary>
+internal sealed class AggregateTagger<TTag> : ITagger<TTag>, IDisposable
+    where TTag : ITag
+{
+    private readonly ImmutableArray<ITagger<TTag>> _taggers;
+
+    public AggregateTagger(ImmutableArray<ITagger<TTag>> taggers)
+        => _taggers = taggers;
+
+    public void Dispose()
+    {
+        foreach (var tagger in _taggers)
+            (tagger as IDisposable)?.Dispose();
+    }
+
+    public event EventHandler<SnapshotSpanEventArgs> TagsChanged
+    {
+        add
+        {
+            foreach (var tagger in _taggers)
+                tagger.TagsChanged += value;
+        }
+
+        remove
+        {
+            foreach (var tagger in _taggers)
+                tagger.TagsChanged -= value;
+        }
+    }
+
+    public IEnumerable<ITagSpan<TTag>> GetTags(NormalizedSnapshotSpanCollection spans)
+    {
+        using var _ = ArrayBuilder<ITagSpan<TTag>>.GetInstance(out var result);
+
+        foreach (var tagger in _taggers)
+            result.AddRange(tagger.GetTags(spans));
+
+        return result.ToImmutable();
```

Comment:
>ToImmutable [](http://example.com/codeflow?start=22&length=11)

I'm curious whether there is any downside to calling ToImmutableAndFree in situations like this (as it has the MoveToImmutable potential advantage).

Reply:
well we don't want to call ToImmtuableAndFree (that would double free).  ToImmutableAndClear would be slightly more efficient.  But my general position is that i don't care about those for *top level* rare allocations.  e.g. there's no tight loop asking fort he tagger for a tag-provider.  So it's not critical to avoid any waste here.

Reply:
Fair enough, I'm just trying to map in my head the proper efficient technique to use this.

Reply:
afaict, the 99.9% case should be ToImmutableAndClear.  It's practically what is wanted in all cases.  Either the immutable array is created, and the builder is freed, or the immutable array is created, and you want to go back to empty in the builder for something like another loop of some algorithm.

ToImmutableAndClear will also be at least as efficient as ToImmutable in all cases, and def more efficient in the case where the array size matches element count.

I would be fine with bulk going and making everything use that in the IDE layer.